### PR TITLE
#8066 Handle TABLE data type in AA api responses

### LINKED ISSUE #8066: Handle new response formats in the AA Run Bot and Run API Task bricks
#### `TABLE`:
```
[
  {
    "Date": "2024-03-25",
    "Items": "Pepperoni Pizza, Soda",
    "Total": "20.00"
  },
  {
    "Date": "2023-09-03",
    "Items": "Steak, Baked Potato, Iced Tea",
    "Total": "42.00"
  },
  {
    "Date": "2023-09-03",
    "Items": "Sandwich, Chips, Soda",
    "Total": "10.25"
  }
]
```
* An array of row objects with each row object having Col:Value pairs

### PR DESCRIPTION
**Pull request by BLoe**
## What does this PR do?

- Closes #8066 

## Checklist

- [x] Add tests and/or storybook stories
- [x] Designate a primary reviewer - @grahamlangford

### CODE REVIEW

File: src/contrib/automationanywhere/contract.ts [hunk 1]
```diff
@@ -128,6 +144,8 @@ export type OutputValue = {
     key: string;
     value: OutputValue;
   }>;
+  // Seems like table property is optional on the response value objects
+  table?: TableValue;
```
Comment (twschiller) [inline src/contrib/automationanywhere/contract.ts:148]: I think their output might only have the type that's recorded under `type`? Or are they other keys always on there?

File: src/contrib/automationanywhere/contract.ts [hunk 2]
```diff
@@ -119,6 +119,22 @@ export const FAILURE_STATUSES = new Set([
   "RUN_TIMED_OUT",
 ]);
 
+type TableColumnSchema = {
```
Comment (twschiller) [inline src/contrib/automationanywhere/contract.ts:122]: Did you use any AA reference docs that you might link to? Or did we just have to reverse engineer the type from the API call?

File: src/contrib/automationanywhere/aaUtils.ts [hunk 3]
```diff
@@ -165,8 +186,17 @@ function mapBotOutput(value: OutputValue): Primitive | UnknownObject {
...
+      if (value.table == null) {
+        return [];
+      }
+
+      return mapBotTableValue(value.table);
+    }
+
     default: {
       const exhaustiveCheck: never = value.type;
+      console.log("*** TABLE:", value);
```
Comment (twschiller) [inline src/contrib/automationanywhere/aaUtils.ts:199]: Remove logging statement

File: src/contrib/automationanywhere/aaUtils.ts [hunk 4]
```diff
@@ -145,7 +146,27 @@ export function mapBotInput(data: UnknownObject) {
   });
 }
 
-function mapBotOutput(value: OutputValue): Primitive | UnknownObject {
+function mapBotTableValue(tableValue: TableValue): UnknownObject[] {
+  const { schema, rows } = tableValue;
+  return rows.map((row) => {
+    const rowObject: UnknownObject = {};
+    for (const [index, value] of row.values.entries()) {
```
Comment (twschiller) [inline src/contrib/automationanywhere/aaUtils.ts:153]: Is this a place where the 2 arrays should be zipped together?

File: src/contrib/automationanywhere/aaUtils.ts [hunk 5]
```diff
@@ -145,7 +146,27 @@ export function mapBotInput(data: UnknownObject) {
...
+      // eslint-disable-next-line security/detect-object-injection -- Using index lookups
+      const columnSchema = schema[index];
+      if (columnSchema == null) {
```
Comment (twschiller) [inline src/contrib/automationanywhere/aaUtils.ts:156]: In what situation would the schema be missing? Is this to guard against arrays of different lengths?

**Review by twschiller — APPROVED** [review]
@BLoe generally looking good - left some clarification questions/comments

File: src/contrib/automationanywhere/aaUtils.test.ts [hunk 6]
```diff
@@ -204,6 +307,23 @@ describe("selectBotOutput", () => {
         baz: true,
         qux: 42,
       },
+      out_Table: [
```
Comment (twschiller) [inline src/contrib/automationanywhere/aaUtils.test.ts:310]: 💯

File: src/contrib/automationanywhere/contract.ts (hunk 1 above)
Comment (BLoe) [reply src/contrib/automationanywhere/contract.ts:148]: No, they always have a bunch of extra fields with empty values

File: src/contrib/automationanywhere/contract.ts (hunk 2 above)
Comment (BLoe) [reply src/contrib/automationanywhere/contract.ts:122]: I just reverse engineered it from the api call

File: src/contrib/automationanywhere/aaUtils.ts (hunk 4 above)
Comment (BLoe) [reply src/contrib/automationanywhere/aaUtils.ts]: Yep, `zipWith()` is a little cleaner, thanks!

File: src/contrib/automationanywhere/contract.ts (hunk 2 above)
Comment (BLoe) [reply src/contrib/automationanywhere/contract.ts:122]: I'll add a comment

**Pull request merged**